### PR TITLE
PoC : A slim app component CRD to drive build & deployments from source

### PR DESCRIPTION
**Pull request by sbose78**
Using the following information, I will be deploying a `component` called `f8-ui` which belongs to the app `openshiftio` :  
- build type
- git url ( codebase )
- port on which the container app is listening on

The OpenShift DevOps console should be able to hide out the complexity of kubernetes from the user.

This PoC shows that using the new CRD controller that watches for new `component` custom resources, one could have an running container from a Github source. 

Note: The controller/watch only *creates* if absent. Doesn't support *patch* yet.

### Create the CRD:
https://github.com/redhat-developer/devopsconsole-operator/blob/1e680698ae0ab5f6d64ad8fca6e1d27d9e12f539/deploy/crds/devopsconsole_v1alpha1_component_crd.yaml

```
oc create -f deploy/crds/devopsconsole_v1alpha1_component_crd.yaml
```
You could verify by doing an `oc get crds`.

Run the controller locally by doing a 
```
$ operator-sdk up local
```
### Create a custom resource 

```
oc create -f examples/devopsconsole_v1alpha1_component_cr.yaml
```
https://github.com/redhat-developer/devopsconsole-operator/blob/cd708f679d06196efedcda98b0f7cdae00f92f3e/examples/devopsconsole_v1alpha1_component_cr.yaml

Alternatively, you could 

### Call the REST API to create the component object

( which is what the UI would be doing ) 

```
POST /apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions/components.devopsconsole.openshift.io

apiVersion: devopsconsole.openshift.io/v1alpha1
kind: Component
metadata:
  name: f8-ui
spec:
  app: "openshiftio"
  buildtype: "nodejs"
  codebase: "https://github.com/sbose78/nodejs-rest-http-crud"
  listenport: 8080
```

Note: The information that is used here to create these objects are 
- build type
- git url ( codebase )
- port on which the container app is listening on

### Outcome
This should trigger the controller `watch` to create the 'primary resource' which is `f8-ui` of type `component`

```
$ oc get components
NAME      AGE
f8-ui     4h
```
... and the secondary resources are the following. 
Not sure if it would be right to call the following as secondary resources though:
- pod/f8-ui-buildconfig-1-build ( seems to be secondary to the build object )
- replicationcontroller/f8-ui-1  ( seems to be secondary to the deploymentConfig object )

```
$ oc get all
NAME                            READY     STATUS      RESTARTS   AGE
pod/f8-ui-1-d6bv8               1/1       Running     0          4h
pod/f8-ui-buildconfig-1-build   0/1       Completed   0          4h

NAME                            DESIRED   CURRENT   READY     AGE
replicationcontroller/f8-ui-1   1         1         1         4h

NAME                                       REVISION   DESIRED   CURRENT   TRIGGERED BY
deploymentconfig.apps.openshift.io/f8-ui   1          1         1         config

NAME                                               TYPE      FROM         LATEST
buildconfig.build.openshift.io/f8-ui-buildconfig   Source    Git@master   1

NAME                                           TYPE      FROM          STATUS     STARTED       DURATION
build.build.openshift.io/f8-ui-buildconfig-1   Source    Git@a4743fd   Complete   5 hours ago   35s

NAME                                   DOCKER REPO                          TAGS      UPDATED
imagestream.image.openshift.io/f8-ui   172.30.1.1:5000/consolespace/f8-ui   latest    5 hours ago
```

( The build & deployment needs to be manually triggered at the moment )

### Labelling
The `app` label groups the relevant objects together as you can see in :

screen shot 2019-02-20 at 2 53 33 pm

### What next ?

- Do we need to have a `codebase` CRD ? Instead of that, could we use the `sourceSecret` and maintain the map between a git url and credential in the app `component` CR ? https://jira.coreos.com/browse/ODC-65#add-comment
- Should we evaluate & use https://github.com/snowdrop/component-operator/
- Come up with the `finalizers` and general watch strategies when secondary resources change.
- Define how to associate services from the developer catalog.

### CODE REVIEW

File: pkg/controller/component/component_controller.go [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+
+		reqLogger.Info("Creating a new BuildConfig", "Buildconfig.Namespace", newBuildConfig.Namespace, "BuildConfig.Name", newBuildConfig.Name)
+
+		err = r.client.Create(context.TODO(), newBuildConfig)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// BuildConfig created successfully - don't requeue
+		reqLogger.Info("BuildConfig created", "BuildConfig.Namespace", foundBuildConfig.Namespace, "BuildConfig.Name", foundBuildConfig.Name)
```
[the remark this answers is not in the document]
Comment (sbose78) [reply pkg/controller/component/component_controller.go]: Yes!

**Comment by sbose78** [comment]
Feedback: ensure namespace info is present in the CR object so that objects returned as part of `oc get components --all-namespaces=true` has the namespace metadata.

**Comment by sbose78** [comment]
Install olm 

- https://github.com/operator-framework/operator-lifecycle-manager/blob/master/Documentation/install/install.md
- Verify:  `oc get clusterserviceversions` should be a valid custom resource type.

Generate & Write the manifest

```
$ operator-sdk olm-catalog gen-csv --csv-version 0.1.0 --update-crds
$ operator-sdk build docker.io/sbose78/devopsconsole-operator.0.1.0
$ docker push ...
``` 

Modify `deploy/olm-catalog/devopsconsole-operator.csv.yaml` with the correct image url ( and missing fields ) and then apply it 

```
$ oc apply -f deploy/olm-catalog/devopsconsole-operator.csv.yaml
$ oc get clusterserviceversion
NAME                            DISPLAY                             VERSION   REPLACES   PHASE
devopsconsole-operator.v0.1.0   OpenShift DevOps Console Operator   0.1.0                
```
screen shot 2019-02-26 at 2 37 43 pm

**Comment by cmoulliard** [comment]
Why do you use another CRD instead of the k8s Application CRD [1] ? @sbose78 

[1] https://github.com/kubernetes-sigs/application/

**Comment by cmoulliard** [comment]
This PR will only support the `outerloop` strategy (= s2i build) and not at all the `innerloop` [1] which is needed at the beginning of the process when the developer will start to play with an Application deployed on OpenShift. Before to continue the development, I think that it could be interesting that we discuss @sbose78

[1] https://github.com/snowdrop/component-operator/blob/master/pkg/pipeline/innerloop/install.go#L60

**Comment by sbose78** [comment]
Agreed @cmoulliard

File: pkg/apis/devopsconsole/v1alpha1/component_types.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+// ComponentSpec defines the desired state of Component
+// +k8s:openapi-gen=true
+type ComponentSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
+	// Add custom validation using kubebuilder tags: https://book.kubebuilder.io/beyond_basics/generating_crd.html
+	App        string `json:"app"`
+	Codebase   string `json:"codebase"`
+	Buildtype  string `json:"nodejs"`
```
Comment (akashshinde) [inline pkg/apis/devopsconsole/v1alpha1/component_types.go]: @sbose78  Shouldn't It supposed to be `build_type` ?

Comment (sbose78) [reply pkg/apis/devopsconsole/v1alpha1/component_types.go]: Yes, right!
In your devfile PoC , change it and use this branch :)

Comment (sbose78) [reply pkg/apis/devopsconsole/v1alpha1/component_types.go]: Yes, right!
In your devfile PoC , change it and use this branch :)

File: pkg/controller/component/component_controller.go [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+	// Watch for changes to primary resource Component
+	err = c.Watch(&source.Kind{Type: &devopsconsolev1alpha1.Component{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	/*
+
+		// TODO(user): Modify this to be the types you create that are owned by the primary resource
+		// Watch for changes to secondary resource Pods and requeue the owner Component
```
Comment (jarifibrahim) [inline pkg/controller/component/component_controller.go]: @sbose78 We're not watching any secondary resources. Is this intentional?

File: pkg/apis/devopsconsole/v1alpha1/component_types.go (hunk 2 above)
Comment (cmoulliard) [reply pkg/apis/devopsconsole/v1alpha1/component_types.go]: Is there a reason why you don't use our [ComponentSpec](https://github.com/snowdrop/component-operator/blob/master/pkg/apis/component/v1alpha1/component_types.go) enriched with your needs ? @sbose78

Comment (sbose78) [reply pkg/apis/devopsconsole/v1alpha1/component_types.go]: @cmoulliard , Corinne is on that - she is evaluating it and getting something working. This is a PoC PR.

File: pkg/controller/component/component_controller.go (hunk 3 above)
Comment (sbose78) [reply pkg/controller/component/component_controller.go]: Just for simplicity here. But we should watch changes to BuildConfig and sync them back.

**Comment by corinnekrych** [comment]
@sbose78 once we've got https://github.com/redhat-developer/devconsole-operator/pull/44 merged maybe it's time to close that one :)

**Comment by baijum** [comment]
> @sbose78 once we've got #44 merged maybe it's time to close that one :)

I am closing this for now!

**Pull request closed without merging**